# Advent of Code 2020 Day 2: Password Philosophy

## Import libraries

In [1]:
import pandas as pd
pd.set_option('display.max_rows', None)

## Part 1

### Description Part 1

Your flight departs in a few days from the coastal airport; the easiest way down to the coast from here is via toboggan.

The shopkeeper at the North Pole Toboggan Rental Shop is having a bad day. "Something's wrong with our computers; we can't log in!" You ask if you can take a look.

Their password database seems to be a little corrupted: some of the passwords wouldn't have been allowed by the Official Toboggan Corporate Policy that was in effect when they were chosen.

To try to debug the problem, they have created a list (your puzzle input) of passwords (according to the corrupted database) and the corporate policy when that password was set.

For example, suppose you have the following list:

1-3 a: abcde
1-3 b: cdefg
2-9 c: ccccccccc
Each line gives the password policy and then the password. The password policy indicates the lowest and highest number of times a given letter must appear for the password to be valid. For example, 1-3 a means that the password must contain a at least 1 time and at most 3 times.

In the above example, 2 passwords are valid. The middle password, cdefg, is not; it contains no instances of b, but needs at least 1. The first and third passwords are valid: they contain one a or nine c, both within the limits of their respective policies.

How many passwords are valid according to their policies?

### Import data

In [2]:
dfa1 = pd.read_csv('day_2_input.csv', header = None)

In [3]:
dfa1.head()

0
0          5-6 s: zssmssbsms
1            3-6 j: jjjjjrrj
2           4-7 k: kfkgkkkkk
3             2-3 n: nkbgfnn
4  7-12 h: hhhhhhdhhhhhfhhhh

### Separate into columns

In [4]:
# Name original column
dfa1.columns = ['original']

In [5]:
dfa1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 1 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   original  1000 non-null   object
dtypes: object(1)
memory usage: 7.9+ KB


In [6]:
# Pop out n-n into separate column
dfa1['n1'] = dfa1['original'].apply(lambda x: x[x.find(x, 0) : x.find('-')])
dfa1['n2'] = dfa1['original'].apply(lambda x: x[x.find('-') + 1 : x.find(' ')])
dfa1['letter'] = dfa1['original'].apply(lambda x: x[x.find(' ') + 1 : x.find(':')])
dfa1['password'] = dfa1['original'].apply(lambda x: x[x.find(': ') + 1 :])

In [7]:
dfa1['n1'] = dfa1['n1'].str.strip()
dfa1['n2'] = dfa1['n2'].str.strip()
dfa1['letter'] = dfa1['letter'].str.strip()
dfa1['password'] = dfa1['password'].str.strip()

In [8]:
check = dfa1[dfa1['password'].str.contains(' ')]

print(check)

Empty DataFrame
Columns: [original, n1, n2, letter, password]
Index: []


In [9]:
dfa1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   original  1000 non-null   object
 1   n1        1000 non-null   object
 2   n2        1000 non-null   object
 3   letter    1000 non-null   object
 4   password  1000 non-null   object
dtypes: object(5)
memory usage: 39.2+ KB


In [10]:
dfa1['n1'] = dfa1['n1'].astype('int')
dfa1['n2'] = dfa1['n2'].astype('int')

In [11]:
dfa1['letter_count'] = dfa1.apply(lambda x: x['password'].count(x['letter']), axis = 1)

In [12]:
dfa1.head()

original  n1  n2 letter           password  letter_count
0          5-6 s: zssmssbsms   5   6      s         zssmssbsms             6
1            3-6 j: jjjjjrrj   3   6      j           jjjjjrrj             6
2           4-7 k: kfkgkkkkk   4   7      k          kfkgkkkkk             7
3             2-3 n: nkbgfnn   2   3      n            nkbgfnn             3
4  7-12 h: hhhhhhdhhhhhfhhhh   7  12      h  hhhhhhdhhhhhfhhhh            15

In [13]:
dfa2 = dfa1[(dfa1['letter_count'] <= dfa1['n2']) & (dfa1['letter_count'] >= dfa1['n1'])]

In [14]:
dfa2.shape

(638, 6)

In [15]:
check1 = dfa2[dfa2['letter_count'] < dfa2['n1']]
check2 = dfa2[dfa2['letter_count'] > dfa2['n2']]

print(check1, check2)

Empty DataFrame
Columns: [original, n1, n2, letter, password, letter_count]
Index: [] Empty DataFrame
Columns: [original, n1, n2, letter, password, letter_count]
Index: []


## Solution: 638 passwords are valid according to the above policies

## Part 2

### Description Part 2

While it appears you validated the passwords correctly, they don't seem to be what the Official Toboggan Corporate Authentication System is expecting.

The shopkeeper suddenly realizes that he just accidentally explained the password policy rules from his old job at the sled rental place down the street! The Official Toboggan Corporate Policy actually works a little differently.

Each policy actually describes two positions in the password, where 1 means the first character, 2 means the second character, and so on. (Be careful; Toboggan Corporate Policies have no concept of "index zero"!) Exactly one of these positions must contain the given letter. Other occurrences of the letter are irrelevant for the purposes of policy enforcement.

Given the same example list from above:

1-3 a: abcde is valid: position 1 contains a and position 3 does not.
1-3 b: cdefg is invalid: neither position 1 nor position 3 contains b.
2-9 c: ccccccccc is invalid: both position 2 and position 9 contain c.
How many passwords are valid according to the new interpretation of the policies?

In [16]:
dfa1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   original      1000 non-null   object
 1   n1            1000 non-null   int32 
 2   n2            1000 non-null   int32 
 3   letter        1000 non-null   object
 4   password      1000 non-null   object
 5   letter_count  1000 non-null   int64 
dtypes: int32(2), int64(1), object(3)
memory usage: 39.2+ KB


In [17]:
dfa1.head()

original  n1  n2 letter           password  letter_count
0          5-6 s: zssmssbsms   5   6      s         zssmssbsms             6
1            3-6 j: jjjjjrrj   3   6      j           jjjjjrrj             6
2           4-7 k: kfkgkkkkk   4   7      k          kfkgkkkkk             7
3             2-3 n: nkbgfnn   2   3      n            nkbgfnn             3
4  7-12 h: hhhhhhdhhhhhfhhhh   7  12      h  hhhhhhdhhhhhfhhhh            15

In [18]:
dfa3 = dfa1.drop('letter_count', axis = 1)

In [19]:
dfa3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   original  1000 non-null   object
 1   n1        1000 non-null   int32 
 2   n2        1000 non-null   int32 
 3   letter    1000 non-null   object
 4   password  1000 non-null   object
dtypes: int32(2), object(3)
memory usage: 31.4+ KB


In [20]:
dfa3['n11'] = dfa3['n1'] - 1
dfa3['n22'] = dfa3['n2'] - 1

In [21]:
dfa3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   original  1000 non-null   object
 1   n1        1000 non-null   int32 
 2   n2        1000 non-null   int32 
 3   letter    1000 non-null   object
 4   password  1000 non-null   object
 5   n11       1000 non-null   int32 
 6   n22       1000 non-null   int32 
dtypes: int32(4), object(3)
memory usage: 39.2+ KB


In [22]:
dfa3['n1_letter'] = dfa3.apply(lambda x: x['password'][x['n11']], axis=1)
dfa3['n2_letter'] = dfa3.apply(lambda x: x['password'][x['n22']], axis=1)

In [23]:
dfa3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 9 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   original   1000 non-null   object
 1   n1         1000 non-null   int32 
 2   n2         1000 non-null   int32 
 3   letter     1000 non-null   object
 4   password   1000 non-null   object
 5   n11        1000 non-null   int32 
 6   n22        1000 non-null   int32 
 7   n1_letter  1000 non-null   object
 8   n2_letter  1000 non-null   object
dtypes: int32(4), object(5)
memory usage: 54.8+ KB


In [24]:
dfa4 = dfa3[(dfa3['n1_letter'] == dfa3['letter']) | (dfa3['n2_letter'] == dfa3['letter'])]

In [25]:
dfa5 = dfa4[dfa4['n1_letter'] != dfa4['n2_letter']]

In [26]:
dfa5.shape

(699, 9)

## Solution: 699 passwords are valid according to the above policies